go/arrow/compute/internal/kernels/vector_run_end_encode.go

```diff
@@ -0,0 +1,953 @@
+		return false
+	}
+
+	*out = bitutil.BitIsSet(inputValues, int(offset))
+	return true
+}
+
+func writeBoolVal(result *exec.ExecResult, offset int64, valid bool, value bool) {
+	if len(result.Buffers[0].Buf) != 0 {
+		bitutil.SetBitTo(result.Buffers[0].Buf, int(offset), valid)
+	}
+	bitutil.SetBitTo(result.Buffers[1].Buf, int(offset), value)
+}
+
+type runEndEncodeLoopFixedWidth[R RunEndsType, V exec.FixedWidthTypes | bool] struct {
+	inputLen, inputOffset int64
+	inputValidity         []byte
+	inputValues           []byte
+	valueType             arrow.DataType
+
+	readValue  func(inputValidity, inputValues []byte, offset int64, out *V) bool
+	writeValue func(*exec.ExecResult, int64, bool, V)
+}
+
+func (re *runEndEncodeLoopFixedWidth[R, V]) WriteEncodedRuns(out *exec.ExecResult) int64 {
+	outputRunEnds := exec.GetData[R](out.Children[0].Buffers[1].Buf)
+
+	readOffset := re.inputOffset
+	var currentRun V
+	curRunValid := re.readValue(re.inputValidity, re.inputValues, readOffset, &currentRun)
+	readOffset++
+
+	var writeOffset int64
+	for readOffset < re.inputOffset+re.inputLen {
+		var value V
+		valid := re.readValue(re.inputValidity, re.inputValues, readOffset, &value)
+		if valid != curRunValid || value != currentRun {
+			// close the current run by writing it out
+			re.writeValue(&out.Children[1], writeOffset, curRunValid, currentRun)
+			runEnd := R(readOffset - re.inputOffset)
+			outputRunEnds[writeOffset] = runEnd
+			writeOffset++
+			curRunValid, currentRun = valid, value
+		}
+		readOffset++
+	}
+
+	re.writeValue(&out.Children[1], writeOffset, curRunValid, currentRun)
+	outputRunEnds[writeOffset] = R(re.inputLen)
+	return writeOffset + 1
+}
+
+func (re *runEndEncodeLoopFixedWidth[R, V]) CountNumberOfRuns() (numValid, numOutput int64) {
+	offset := re.inputOffset
+	var currentRun V
+	curRunValid := re.readValue(re.inputValidity, re.inputValues, offset, &currentRun)
+	offset++
+
+	if curRunValid {
+		numValid = 1
+	}
+	numOutput = 1
+
+	for offset < re.inputOffset+re.inputLen {
+		var value V
+		valid := re.readValue(re.inputValidity, re.inputValues, offset, &value)
+		offset++
+		// new run
+		if valid != curRunValid || value != currentRun {
+			currentRun = value
+			curRunValid = valid
+
+			numOutput++
+			if valid {
+				numValid++
+			}
+		}
+	}
+	return
+}
+
+func (re *runEndEncodeLoopFixedWidth[R, V]) PreallocOutput(ctx *exec.KernelCtx, numOutput int64, out *exec.ExecResult) {
+	runEndsBuffer := ctx.Allocate(int(numOutput) * int(SizeOf[R]()))
+	var validityBuffer *memory.Buffer
+	if len(re.inputValidity) > 0 {
+		validityBuffer = ctx.AllocateBitmap(numOutput)
+	}
+
+	var valueBuffer *memory.Buffer
+	bufSpec := re.valueType.Layout().Buffers[1]
+	if bufSpec.Kind == arrow.KindBitmap {
+		valueBuffer = ctx.AllocateBitmap(numOutput)
+	} else {
+		valueBuffer = ctx.Allocate(int(numOutput) * bufSpec.ByteWidth)
+	}
+
+	reeType := arrow.RunEndEncodedOf(exec.GetDataType[R](), re.valueType)
+	out.Release()
```

Comment:
Go question: re-assigning to `out` doesn't automatically Release it?

Reply:
it doesn't call the `Release()` method. Go doesn't have destructors.

Reply:
it's a case of wanting to propagate. Honestly, this is *technically* unnecessary since I have the kernels set to not preallocate anything, but I'm doing this to make sure that a future change doesn't accidentally cause a memory leak or something.